### ISSUE TITLE
IDE0100 (Remove unnecessary equality operator) catches `x == false` as redundant

### ISSUE THREAD

**Issue by bzd3y** [dotnet/roslyn#73787]
**Version Used**: 
VS 17.10.1

This behavior seems to be new in 17.10(.1?).

**Steps to Reproduce**:

Add the following code to a program:

```csharp
	bool x = true;

	if (x == false)
	{
		// do something...
	}
```

The above if statement is not redundant. It could arguably be _simplified_ by using an `!` as it suggests as a fix, but that should be a different rule or at least configurable for this rule. `!` is almost always not as readable.

**Diagnostic Id**: IDE0100 (Remove unnecessary equality operator)

**Expected Behavior**: The analyzer would not flag the above `if` for being redundant.

**Actual Behavior**: It does and the fix is to refactor to use `!`.

**Comment by CyrusNajmabadi** [comment]
This is by design. The analyzer flags Boolean expressions that have a shorter, idiomatic, language form. `!` for `==false` is one of those forms and was intentionally picked. 

If you don't want a particular item changed, you can suppress the analyzer at that location, file, or disable it altogether.

**Comment by bzd3y** [comment]
It seems a little hasty to close this with no discussion. This was not the case pre 17.10.x. The statements being flagged now were not flagged in 17.9.7. So because of a design change my code suddenly has hundreds of suggestions.

So this is effectively a breaking change. It would definitely be one if somebody chose to flag these as errors and not suggestions. The fact that it can be suppressed or disabled doesn't change that. The same is true for any other breaking change where the proposed solution could just be to not use whatever was breaking.

I don't want to go through hundreds of places and suppress this suggestion, not just because of how tedious it might be, but because it would make the code extremely messy. And disabling it isn't reasonable either. I want to keep this rule to flag actual redundant equalities.

I'm sure it seems like I am arguing, but I think this is valid push back.

Again, this is *not* a redundant equality. It is a check if the value **is** `false` and **not** a check if the value **is not** `true`.

And speaking of "is", it doesn't flag for `x is false`, even though that seems just as redundant. And I am again wondering if the equality operators are now just obsolete for constants/literals because while they are syntactically valid, I can't configure the compiler to not complain about them without configuring it to not complain about anything.

So would it make more sense for me to create an analyzer feature request issue? And is there a chance that would even be considered?

**Comment by chrisoverzero** [comment]
> It is a check if the value is `false` and **not** a check if the value **is not** `true`.

For a value of type `bool`, how is that different?

**Comment by bzd3y** [comment]
> For a value of type bool, how is that different?

Sorry, that probably wasn't very clear, but good question. For all intents and purposes here, at least, it isn't. My point is that `!` is "not true", which is what is actually redundant, if anything, to "equals false". It is shorthand for `!= true`, which is the redundant equality. `x == false` just isn't redundant. That would imply that `false` itself is redundant because you can do `!true`. As an example of what I mean, if you make a truth table, or look at one somewhere, it is generally going to have `true` and `false`, NOT `true and `not true`. `false` IS the definitive value. Any alternative is redundant to it.

And part of my reason for stating it that way was to lead into the fact that the analyzer doesn't flag `x is false` (or `x is not true`). 

And maybe this is all academic (it is). This is less about a Lilliputian argument about `!` vs. `== false` and more about it being configurable, like most of the other analyzer rules.

There seems to be a preference for not using `==` or `false` and certainly never at the same time. And if some people prefer that, then that is fine. But if it was configurable then it could provide value to everybody.

**Comment by bzd3y** [comment]
Bumping this. It seems like there needs to be more discussion on this.

**Comment by richardissimo** [comment]
Here is another [example where someone wrote a similar analyzer rule](https://community.sonarsource.com/t/boolean-literals-should-not-be-redundant-c-csharpsquid-s1125/19115), which provoked a similar response. After some debate, the author wrote "_Seems to me that this rule is dividing the community. It relies on personal preferences and some like it and some just don`t. Your suggestion of splitting the rule is nice, but at this moment I am not sure it is worth investing time to split it._" 

The "==false" variant is not a redundancy, and the decision to use "!" should be a styling preference (via a different rule).

**Comment by bzd3y** [comment]
> The "==false" variant is not a redundancy, and the decision to use "!" should be a styling preference (via a different rule).

Or through configuration. Most of these rules are configurable to allow for styling/other preferences. There is no reason this one couldn't be.

**Comment by adae-palfinger** [comment]
I'm am in the same boat and in my case warns are reported as error so now I have many places broken where I do not want to do `!`, the `is false` was a style choice and it is not redundant. Changing the style is also not an option, either make this configurable or revert to the original implementation
